### PR TITLE
[mono][NativeAOT] System.Globalization.Native build improvements

### LINKED ISSUE #82389: Fix NativeAOT ICU build integration for Apple mobile platforms
### Description

Currently `System.Globalization.Native` is not being built as part of the build process for targeting Apple mobile platforms with NativeAOT. This implies that applications have to use invariant globalization mode.

### Proposal

Support building `System.Globalization.Native` with NativeAOT when targeting mobile platforms and adjust build integration targets according to the fix:

https://github.com/dotnet/runtime/blob/06dd1262479775558b7b8ce4766ddc0e8ae41214/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets#L67

Until this is fixed, make sure that applications set: `InvariantGlobalization=true`

### PR DESCRIPTION
**Pull request by filipnavara**
Fixes #82389

- Fixes build of System.Globalization.Native for iOS-like, browser, and wasi platforms
- Removes ICU shim from static builds of Mono runtime and builds it as part of libs.native subset
- Switches browser/wasi to use the static libraries

### CODE REVIEW

File: src/native/libs/System.Globalization.Native/CMakeLists.txt
Comment (filipnavara) [inline src/native/libs/System.Globalization.Native/CMakeLists.txt:74]:
```suggestion
```

File: src/native/libs/System.Globalization.Native/pal_collation.c
Comment (filipnavara) [inline src/native/libs/System.Globalization.Native/pal_collation.c:341]: The compilation uses Homebrew ICU headers which are newer (>=71) than the shipped ICU.

**Comment by filipnavara** [comment]
The build failures are relevant. I'll try to remove the duplicate build from Mono sources.

**Comment by filipnavara** [comment]
This fundamentally broken at the moment. We would need to link to the packages obtained from Microsoft.NETCore.Runtime.ICU.Transport. I'll have to rethink how to do that.

**Comment by filipnavara** [comment]
The WASI part has overlap with #82444.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets
Comment (am11) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:66]: @filipnavara, in case it wasn't obvious: the `StaticICULinking` is a bit different. It is meant to run cmake _during_ `dotnet publish -p:PublishAot=true` on user machine. I'm not sure how the variant added in `src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props` will be used here, but we seem to have two other options:

1. either bring iOS to the same plan as existing platforms (run cmake to build Globalization shim during dotnet-publish on user machine).
2. or use it from the framework directory; change line 89 as:
    ```diff
    -       <NativeLibrary Include="$(IntermediateOutputPath)libs/System.Globalization.Native/build/libSystem.Globalization.Native.a" />
    +       <NativeLibrary Include="$(IntermediateOutputPath)libs/System.Globalization.Native/build/libSystem.Globalization.Native.a" Condition="'$(_IsiOSLikePlatform)' != 'true'" />
    +       <NativeLibrary Include="$(IlcFrameworkNativePath)libSystem.Globalization.Native.a" Condition="'$(_IsiOSLikePlatform)' == 'true'" />
    ```
      * Note that we already have (an unused!) copy of `libSystem.Globalization.Native.a` in the framework directory:
          ```zsh
          % ls ~/.nuget/packages/runtime.linux-arm64.microsoft.dotnet.ilcompiler/8.0.0-preview.2.23116.1/framework/*Globalization*.a
          /Users/am11/.nuget/packages/runtime.linux-arm64.microsoft.dotnet.ilcompiler/8.0.0-preview.2.23116.1/framework/libSystem.Globalization.Native.a
          ```
          which comes from https://github.com/dotnet/runtime/blob/af18fd07d1bb5a0d923a49c39b19c6fbcb6dc07b/eng/liveBuilds.targets#L175 (unused files are  missing from the ilc package excludes, which should be cleaned up separately)

I'm not familiar with NativeAOT step in Xamrian publish workflow, but if option 1 is feasible for iOS-like platforms, I'd go for it.

Comment (filipnavara) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:66]: libSystem.Globalization.Native.a was not built for iOS-like platforms, so that's part of the fix.

The final linking is done either through AppleAppBuiler or the Xamarin tooling. Both of them can consume the .a file once it exists.

`StaticICULinking` remains unused for the moment for these platforms.

Comment (am11) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:66]: > `StaticICULinking` remains unused for the moment for these platforms.

I missed that! 😁

> libSystem.Globalization.Native.a was not built for iOS-like platforms, so that's part of the fix.

Cool. For NativeAOT, does Xamarin use shared libraries from `microsoft.dotnet.ilcompiler` package or from the Microsoft.NETCore.App / shared framework? e.g. `libSystem.Globalization.Native.a` _can_ be different if user has csproj file with:

```csproj
<!-- Different version than what comes with SDK https://github.com/dotnet/installer/blob/8ce7bb903a8d818bd288267f82be0d98f5e6c757/src/redist/targets/GenerateBundledVersions.targets#L408 -->
<PackageReference Include="Microsoft.DotNet.ILCompiler" Version="$(DifferentVersion)" />
```

Comment (filipnavara) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:66]: That's kind of unanswered packaging question at the moment. 😅

For the current MonoVM-based workloads it comes from Microsoft.NETCoreApp.Runtime.Mono.<RID> packages. This PR adds libSystem.Globalization.Native.a into the package and removes the same APIs from libmonosgen.a (or whether is the static runtime name).

Comment (ivanpovazan) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:66]: > StaticICULinking remains unused for the moment for these platforms.
Should we add a MSBuild warning/error when someone tries to use it on these platforms?

File: src/native/libs/Common/pal_atomic.h
Comment (pavelsavara) [inline src/native/libs/Common/pal_atomic.h:7]: Browser too, for consistency ?

File: src/native/libs/System.Globalization.Native/CMakeLists.txt
Comment (pavelsavara) [inline src/native/libs/System.Globalization.Native/CMakeLists.txt:3]: Browser ?

File: src/native/libs/System.Globalization.Native/pal_calendarData.c
Comment (pavelsavara) [inline src/native/libs/System.Globalization.Native/pal_calendarData.c:12]: Browser

File: src/libraries/native-binplace.proj
Comment (pavelsavara) [inline src/libraries/native-binplace.proj:27]: is this dealing also with ICU shards ?

File: src/mono/wasm/runtime/CMakeLists.txt
Comment (pavelsavara) [inline src/mono/wasm/runtime/CMakeLists.txt:30]: I must admit I don't understand big picture. Could you please explain the consequences of this for browser ?

File: src/native/libs/Common/pal_atomic.h
Comment (filipnavara) [reply src/native/libs/Common/pal_atomic.h:7]: I think the browser builds currently define `TARGET_UNIX`. I'd expect WASI to do the same but I opted for minimal changes.

File: src/mono/wasm/runtime/CMakeLists.txt
Comment (filipnavara) [reply src/mono/wasm/runtime/CMakeLists.txt:30]: Instead of building the ICU shim as part of the Mono runtime build, it moves the build to libs.native part of the build. Then it links the resulting static library back.

The end result should be roughly identical to the previous status quo.

There are several reasons for this change:
- Reuse of the build system in libs.native instead of maintaining a second copy in Mono build
- iOS-like platforms need the library for both Mono and NativeAOT, so building it as part of Mono runtime is problematic; it can be shared by both runtimes this way
- Browser/WASI are just changed to follow the same pattern to make the build system less complicated

File: src/native/libs/Common/pal_atomic.h
Comment (pavelsavara) [reply src/native/libs/Common/pal_atomic.h:7]: Browser has very good posix emulation, so you are right about `TARGET_UNIX`. I would still prefer more explicit reference.
WASI is not unix.

File: src/mono/wasm/runtime/CMakeLists.txt
Comment (pavelsavara) [reply src/mono/wasm/runtime/CMakeLists.txt:30]: thanks!

Comment (lambdageek) [reply src/mono/wasm/runtime/CMakeLists.txt:30]: /cc @steveisok LibraryBuilder targets should also be aware of this change (although I suspect it doesn't matter to the library builder where the static lib came from, as long as it's there)

File: src/libraries/native-binplace.proj
Comment (pavelsavara) [reply src/libraries/native-binplace.proj:27]: For browser we have this https://github.com/dotnet/runtime/pull/80421 
and for WASI we still need more https://github.com/dotnet/runtime/issues/82567

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets
Comment (filipnavara) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:66]: > Should we add a MSBuild warning/error when someone tries to use it on these platforms?

It would be nice to have explicit check but I would prefer not to mix it into this PR.

I don't touch any of the `StaticICULinking` code here and it's documented to work only on Linux when not cross-compiling:
https://github.com/dotnet/runtime/blob/cabaab5e90e19256ae404970091350bb6f0f7139/src/coreclr/nativeaot/docs/compiling.md?plain=1#L77-L85

File: src/native/libs/Common/pal_atomic.h
Comment (filipnavara) [reply src/native/libs/Common/pal_atomic.h:7]: While I understand your reasoning I hesitate to do the change to reduce collateral damage.

Prior to this PR the ICU shim was built in the Mono runtime and it set `TARGET_UNIX` for both for browser ([here](https://github.com/dotnet/runtime/blob/cabaab5e90e19256ae404970091350bb6f0f7139/src/mono/CMakeLists.txt#L696)) and WASI ([here](https://github.com/dotnet/runtime/blob/cabaab5e90e19256ae404970091350bb6f0f7139/src/mono/CMakeLists.txt#L691)). 

In the new setup the System.Globalization.Native library shares the same build system as System.Native. The existing System.Native build defines TARGET_UNIX for Browser, but not for WASI.

If I add `|| defined(TARGET_BROWSER)` here it will be strictly redundant (since `TARGET_BROWSER` implies `TARGET_UNIX`) and inconsistent with the rest of the System.Native code that has no such checks.

File: src/libraries/native-binplace.proj
Comment (filipnavara) [reply src/libraries/native-binplace.proj:27]: It copies all the icu*.dat files just like mono.proj did.

Comment (filipnavara) [reply src/libraries/native-binplace.proj:27]: On second though, it possibly makes `ICULibNativeFiles` and `ICULibFiles` copying in wasm.proj and wasi.proj redundant.

File: src/native/libs/Common/pal_atomic.h
Comment (akoeplinger) [reply src/native/libs/Common/pal_atomic.h:7]: I'm OK with this difference here for now, though we should probably move away from setting TARGET_UNIX for browser too. For example in the managed src/libraries build we do _not_ set `TargetsUnix` msbuild property which gets used to set `TARGET_UNIX` C# define for neither Browser nor Wasi.

File: src/native/libs/System.Globalization.Native/CMakeLists.txt
Comment (akoeplinger) [inline src/native/libs/System.Globalization.Native/CMakeLists.txt:18]: we should instead wrap the `icu_trace_data` function with `#if defined(ICU_TRACING)`

File: src/native/libs/System.Globalization.Native/CMakeLists.txt
Comment (pavelsavara) [reply src/native/libs/System.Globalization.Native/CMakeLists.txt:69]: Also https://github.com/dotnet/runtime/pull/82444/files#diff-fbc587d0b04feb58ec4c10941c3beb54321a1c10d0ffb09c44ebee0100a58103R21

File: src/native/libs/System.Globalization.Native/CMakeLists.txt
Comment (filipnavara) [reply src/native/libs/System.Globalization.Native/CMakeLists.txt:11]: There's some reserved identifier in the ICU headers (prefixed with `__`). Newer emsdk compiler triggers slightly different warning than the ones that were already suppressed for the same reason.

File: src/native/libs/System.Globalization.Native/CMakeLists.txt
Comment (filipnavara) [reply src/native/libs/System.Globalization.Native/CMakeLists.txt:69]: It always happens somewhere, either at linking the runtime as dynamic library, or as part of the final linking of the application.

In this particular case we happen to build .dylib for iOS which may not be strictly necessary. Since the project is marked as using the "C" language and not "CXX" the C++ standard library is not included implicitly.

We can exclude the .dylib builds as an alternative.

Comment (akoeplinger) [reply src/native/libs/System.Globalization.Native/CMakeLists.txt:69]: Ok, I'm mostly worried about Android where we can't introduce this stdc++ dependency. For Android we're loading ICU .so via dlopen from the system location so we just need to make sure this doesn't apply for Android.

I'd probably move the two `link_libraries` down to the condition where we set STATIC_ICU for iOS/Catalyst etc

File: src/native/libs/System.Globalization.Native/CMakeLists.txt
Comment (akoeplinger) [inline src/native/libs/System.Globalization.Native/CMakeLists.txt:116]: not sure where this will actually find icucore from. I think I'd prefer turning off GEN_SHARED_LIB for iOS/Catalyst etc in general for System.Globalization.Native, it wouldn't do the right thing in the non-static case.

Comment (akoeplinger) [reply src/native/libs/System.Globalization.Native/CMakeLists.txt:116]: or we need to make a .dylib of the "static" version where it links in the icu .a

Comment (filipnavara) [reply src/native/libs/System.Globalization.Native/CMakeLists.txt:116]: This is actually a leftover that I forgot to remove. The branch is never taken on the iOS-like platforms.

File: src/native/libs/System.Globalization.Native/CMakeLists.txt
Comment (filipnavara) [reply src/native/libs/System.Globalization.Native/CMakeLists.txt:69]: On Android you will end up linking to the libmono*.so that has the ICU shim built-in. This PR doesn't affect that. We do build an additional libSystem.Globalization.Native.[a/so] and it likely is not necessary. While not necessary for the supported configurations, it is nice to have for the unsupported NativeAOT/Android builds.

**Comment by kotlarmilos** [comment]
@filipnavara is there anything for this PR or is it waiting for a review?

**Comment by filipnavara** [comment]
> @filipnavara is there anything for this PR or is it waiting for a review?

Waiting for review.

**Review by ivanpovazan — APPROVED** [review]
Minor note: The title seems outdated as the PR changes now affect mono builds and other platforms as well. I don't have a strong opinion but something like `[mono][NativeAOT] System.Globalization.Native build improvements` might be more suitable.

Other than that, LGTM.
Thank you very much!

PS I would leave this to @akoeplinger for final approval.

File: src/mono/wasm/runtime/CMakeLists.txt
Comment (akoeplinger) [reply src/mono/wasm/runtime/CMakeLists.txt:30]: from what I can see the LibraryBuilder doesn't need updating since it just takes all *.a files

**Review by akoeplinger — APPROVED** [review]

**Comment by akoeplinger** [comment]
Failures are unrelated.

**Comment by filipnavara** [comment]
> did anyone check the final post linking binary sizes on this?

Any specific configuration you are concerned about?

For NativeAOT/iOS there's nothing to compare against since it didn't work previously. For everything else it shuffles a bit where/how the globalization shim is built but it doesn't really add or remove any code, or at least not intentionally.

**Comment by steveisok** [comment]
> did anyone check the final post linking binary sizes on this?

Quick spot check on one of Alex's test official builds show no change.  I'll check again when a real one gets through.

**Pull request merged**
